source/fab/tools/compiler.py

```diff
@@ -0,0 +1,301 @@
+
+from fab.tools.categories import Categories
+from fab.tools.flags import Flags
+from fab.tools.tool import VendorTool
+
+
+class Compiler(VendorTool):
+    '''This is the base class for any compiler. It provides flags for
+
+    - compilation only (-c),
+    - naming the output file (-o),
+    - OpenMP
+
+    :param name: name of the compiler.
+    :param exec_name: name of the executable to start.
+    :param vendor: name of the compiler vendor.
+    :param category: the Category (C_COMPILER or FORTRAN_COMPILER).
+    :param compile_flag: the compilation flag to use when only requesting
+        compilation (not linking).
+    :param output_flag: the compilation flag to use to indicate the name
+        of the output file
+    :param omp_flag: the flag to use to enable OpenMP
+    '''
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, name: str, exec_name: str, vendor: str,
+                 category: Categories,
+                 compile_flag: Optional[str] = None,
+                 output_flag: Optional[str] = None,
+                 omp_flag: Optional[str] = None):
+        super().__init__(name, exec_name, vendor, category)
+        self._version = None
+        self._compile_flag = compile_flag if compile_flag else "-c"
+        self._output_flag = output_flag if output_flag else "-o"
+        self._omp_flag = omp_flag
+        self.flags.extend(os.getenv("FFLAGS", "").split())
+
+    def get_hash(self) -> int:
+        ''':returns: a hash based on the compiler name and version.
+        '''
+        return (zlib.crc32(self.name.encode()) +
+                zlib.crc32(str(self.get_version()).encode()))
+
+    def compile_file(self, input_file: Path, output_file: Path,
+                     add_flags: Union[None, List[str]] = None):
+        '''Compiles a file. It will add the flag for compilation-only
+        automatically, as well as the output directives. The current working
+        directory for the command is set to the folder where the source file
+        lives when compile_file is called. This is done to stop the compiler
+        inserting folder information into the mod files, which would cause
+        them to have different checksums depending on where they live.
+
+        :param input_file: the path of the input file.
+        :param outpout_file: the path of the output file.
+        :param add_flags: additional compiler flags.
+        '''
+
+        params: List[Union[Path, str]] = [self._compile_flag]
+        if add_flags:
+            params += add_flags
+
+        params.extend([input_file.name,
+                      self._output_flag, str(output_file)])
+
+        return self.run(cwd=input_file.parent,
+                        additional_parameters=params)
+
+    def check_available(self) -> bool:
+        '''
+        :returns: whether the compiler is available or not. We do
+            this by requesting the compiler version.
+        '''
+        try:
+            version = self.get_version()
```

Comment:
Let me know if you prefer a `which` - it certainly means that we could move more of the very similar code from the various tools into the base class. I think I stuck with a help/version command, since this was used for some tools in the past, so that's how I started :) But I don't mind one way or another.

Reply:
Ah, I forgot one argument that swayed me: since I expect scripts might want to check compiler versions at some stage, using the compiler version as a test will save one call :)    (admittedly, it doesn't matter)